### PR TITLE
Parallelize GBFF computations

### PR DESCRIPTION
**Pull request by pgrinaway**
I'm converting the PyMC model to use arrays inside PyMC objects, instead of arrays of PyMC objects. This is partially finished, but opening the PR now. Once the initial implementation is complete, I will give Celery a try, and perhaps MPI as well. It should be straightforward either way, since the modular nature of the code means only one function will need to change to use a different backend.

### CODE REVIEW

**Comment by jchodera** [comment]
Awesome!

**Comment by pgrinaway** [comment]
Currently running into an issue with vectorizing. I'm wondering if I'm not vectorizing properly? Can I not use an array for the `mu` and `tau` parameters in the `pymc.Normal` object? Here is the traceback in its full glory:

```
  File "parameterize-using-database.py", line 147, in <module>
    obcmodel = model.GBFFThreeParameterModel(database, parameters, utils.array_hydration_energies_factory, gbmodel=2)
  File "/Users/grinawap/gbff_cleanup/gbff/model.py", line 248, in __init__
    super(GBFFThreeParameterModel, self).__init__(database, initial_parameters, parameter_types, hydration_energy_factory)
  File "/Users/grinawap/gbff_cleanup/gbff/model.py", line 35, in __init__
    self.model = self._create_bayesian_gbmodel(solvated_system_database, initial_parameters)
  File "/Users/grinawap/gbff_cleanup/gbff/model.py", line 85, in _create_bayesian_gbmodel
    gbffmodel_with_mols = self._add_parallel_gbffmodel(database, gbffmodel)
  File "/Users/grinawap/gbff_cleanup/gbff/model.py", line 169, in _add_parallel_gbffmodel
    gbffmodel['dg_exp'] = pymc.Normal('dg_exp', mu=np.array(gbffmodel['dg_gbsa']), tau=np.array(gbffmodel['taus']), value = np.array(dg_exp), observed=True)
  File "/Users/grinawap/anaconda/lib/python2.7/site-packages/pymc/distributions.py", line 320, in __init__
    **arg_dict_out)
  File "/Users/grinawap/anaconda/lib/python2.7/site-packages/pymc/PyMCObjects.py", line 762, in __init__
    verbose=verbose)
  File "/Users/grinawap/anaconda/lib/python2.7/site-packages/pymc/Node.py", line 214, in __init__
    Node.__init__(self, doc, name, parents, cache_depth, verbose=verbose)
  File "/Users/grinawap/anaconda/lib/python2.7/site-packages/pymc/Node.py", line 129, in __init__
    self.parents = parents
  File "/Users/grinawap/anaconda/lib/python2.7/site-packages/pymc/Node.py", line 147, in _set_parents
    self.gen_lazy_function()
  File "/Users/grinawap/anaconda/lib/python2.7/site-packages/pymc/PyMCObjects.py", line 811, in gen_lazy_function
    self._logp.force_compute()
  File "LazyFunction.pyx", line 257, in pymc.LazyFunction.LazyFunction.force_compute (pymc/LazyFunction.c:2409)
  File "/Users/grinawap/anaconda/lib/python2.7/site-packages/pymc/distributions.py", line 2979, in wrapper
    return f(value, **kwds)
  File "/Users/grinawap/anaconda/lib/python2.7/site-packages/pymc/distributions.py", line 2170, in normal_like
    return flib.normal(x, mu, tau)
ValueError: setting an array element with a sequence.
```

**Comment by kyleabeauchamp** [comment]
Can you print the values and shape of the parameters in this line?

```
    gbffmodel['dg_exp'] = pymc.Normal('dg_exp', mu=np.array(gbffmodel['dg_gbsa']), tau=np.array(gbffmodel['taus']), value = np.array(dg_exp), observed=True)
```

**Comment by kyleabeauchamp** [comment]
Sounds like a shape error in the arrays or something

**Comment by jchodera** [comment]
Is `mu` an array while `tau` is a scalar?
https://github.com/pgrinaway/gbff/blob/parallel/model.py#L150-L152

Nevermind, I missed some critical details.

**Comment by jchodera** [comment]
Some random ideas:
- Not sure if you can use `np.array(pymc_objects)`.  
- Do you need to provide the `size=N` argument for `pymc.Normal`?

**Comment by kyleabeauchamp** [comment]
I don't think size is necessary for Normal.

I do agree that we want to avoid arrays of pymc objects whenever possible, because (if it works) it calls their wacky container objects.

**Comment by jchodera** [comment]
Avoiding arrays of pymc objects might require more restructuring so that numpy arrays are computed and used throughout the `sigma` > `tau` chain.  It looks like `hydration_energy_function` already returns a numpy array though?

**Comment by pgrinaway** [comment]
Yeah, the `np.array(pymc object)` thing was me messing around--I've removed it (will push the commit in a second). It didn't really seem to make a difference, as I initially didn't use it.

**Comment by pgrinaway** [comment]
I see what you guys are saying. I'm going to try to restructure to avoid making arrays of pymc objects.

**Comment by pgrinaway** [comment]
It seems the problem was actually with the `Quantity` thingies. Sampling works on arrays now. I'm going to clean up all the `print()` calls and some other stuff now,

**Comment by jchodera** [comment]
Awesome!

Does that mean the array of pymc objects passed to Normal works?

**Comment by pgrinaway** [comment]
Actually, I'm not passing arrays of pymc objects anymore. But it might work--I got rid of the arrays of objects before trying the quantity thing

**Comment by pgrinaway** [comment]
To clarify this a bit, each PyMC object has array-valued parameters, but I haven't made an array of PyMC objects.

**Comment by jchodera** [comment]
Great, that sounds the most reasonable!

**Comment by pgrinaway** [comment]
Ok, so currently I have this running on the cluster with 50 worker threads, will report back when that is complete. Otherwise, this seems mostly finished.

**Comment by jchodera** [comment]
Sweet!  
- what did you use for the backend?
- any idea of whether this is giving you the speedup you were looking for?

Sounds like this strategy will work for all of our parameterization needs.

**Comment by pgrinaway** [comment]
- I'm using Celery as the distributed computing framework, which is in turn using redis as its broker and backend (there are other options, but this one seemed the simplest for the applications).
- Right now, the initial simulation is still serial (it modifies dictionaries in place). I will modify that to be parallel as well, but I wanted to leave out that potential issue until I got the main model iterations to be parallel. I'll post here when I see what the performance looks like.

**Comment by pgrinaway** [comment]
Hm, a bit of an issue on the cluster it seems. The simulation ran into `Exception: Error invoking kernel: CUDA_ERROR_INVALID_HANDLE (400)`, and consistently seems to be doing so. Note that I am not using `CUDA` (or at least I'm trying to avoid it). I specifically set the `Platform` to `CPU` now, and even `module unload cuda`. Not sure what is going on; will continue investigating tomorrow.

**Comment by pgrinaway** [comment]
For reference, here is the (fairly polluted) stack trace:

```
  File "/cbio/jclab/projects/pgrinaway/gbff/parameterize-using-database.py", line 150, in <module>
    sampler.isample(iter=mcmcIterations, burn=0, verbose=True, progress_bar=True)
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/Model.py", line 658, in isample
    self.iprompt(out=out)
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/Model.py", line 713, in iprompt
    reraise(a, b, c)
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/Model.py", line 648, in samp_targ
    self.sample(*args, **kwds)
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/MCMC.py", line 279, in sample
    Sampler.sample(self, iter, length, verbose)
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/Model.py", line 251, in sample
    self._loop()
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/MCMC.py", line 315, in _loop
    step_method.step()
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/StepMethods.py", line 511, in step
    logp_p = self.logp_plus_loglike
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/StepMethods.py", line 308, in logp_plus_loglike
    sum = logp_of_set(self.markov_blanket)
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/Node.py", line 39, in logp_of_set
    six.reraise(*exc)
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/Node.py", line 30, in logp_of_set
    logp += obj.logp
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/PyMCObjects.py", line 908, in get_logp
    logp = self._logp.get()
  File "LazyFunction.pyx", line 280, in pymc.LazyFunction.LazyFunction.get (pymc/LazyFunction.c:2554)
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/Container.py", line 541, in get_value
    self.DCValue.run()
  File "Container_values.pyx", line 65, in pymc.Container_values.DCValue.run (pymc/Container_values.c:1226)
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/pymc/PyMCObjects.py", line 468, in get_value
    _value = self._value.get()
  File "LazyFunction.pyx", line 280, in pymc.LazyFunction.LazyFunction.get (pymc/LazyFunction.c:2568)
  File "/cbio/jclab/projects/pgrinaway/gbff/hydration_energies/energytasks.py", line 170, in hydration_energies
    return compute_hydration_energies_celery(database, parameters, platform_name='CPU')
  File "/cbio/jclab/projects/pgrinaway/gbff/hydration_energies/energytasks.py", line 166, in compute_hydration_energies_celery
    return energy_group.get()
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/celery/result.py", line 578, in get
    interval=interval, callback=callback, no_ack=no_ack)
  File "/cbio/jclab/home/pgrinaway/anaconda/lib/python2.7/site-packages/celery/result.py", line 691, in join_native
    raise value
Exception: Error invoking kernel: CUDA_ERROR_INVALID_HANDLE (400)
```

File: utils.py [hunk 1]
```diff
@@ -369,24 +259,25 @@ def compute_hydration_energy(entry, parameters, platform_name="Reference"):
...
-    # Create context for solvent system.
-    timestep = 2.0 * units.femtosecond
-    solvent_integrator = openmm.VerletIntegrator(timestep)
-    solvent_context = openmm.Context(solvent_system, solvent_integrator, platform)
-
-    # Create context for vacuum system.
-    vacuum_integrator = openmm.VerletIntegrator(timestep)
-    vacuum_context = openmm.Context(vacuum_system, vacuum_integrator, platform)
+    solvent_context = openmm.Context(solvent_system, solvent_integrator)
+    vacuum_context = openmm.Context(vacuum_system, vacuum_integrator)
```
Comment (kyleabeauchamp) [inline utils.py]: This is a context creation without an explicit platform being set, which presumably might default to CUDA on some machines.  Could this be it?

Comment (pgrinaway) [inline utils.py]: Possibly. I don't think this function is being called, but I will check.

**Comment by jchodera** [comment]
Did @danielparton have similar issues with Ensembler?

I have probably mentioned this before, but it is also important to make sure the Context objects are deleted for GPUs so they don't eat up all available resources.

**Comment by jchodera** [comment]
Since you're distributing the computation now, I bet you can get rid of `Context` caching and greatly simplify things.

**Comment by pgrinaway** [comment]
I believe @kyleabeauchamp was correct. Actually, the function he pointed out is no longer called (I have no idea why that was causing a problem), so I removed it. The sampling is currently proceeding without error (and without that old function). I don't cache `Context`s anymore either.

**Comment by pgrinaway** [comment]
Also, I do a `del context` after each energy computation, regardless of platform.

**Comment by pgrinaway** [comment]
Ok, something weird is going on with the hdf5 backend on the cluster. I'm currently running it with 100 molecules on all cores of `jak`, with 10,000 iterations.

**Comment by pgrinaway** [comment]
Ok, so it appears Celery works as I expected, which is pretty cool. I started off the calculation with 32 threads on jak, but then started an additional 64 (32 on csk and 32 on egfr), and they were able to join in pretty seamlessly.

**Comment by jchodera** [comment]
Sweet!

The only thing to watch out for in future might be to have some check
against the expected local version number in simtk.openmm.version, just to
make sure the locally installed version of openmm is as expected.

**Comment by pgrinaway** [comment]
> The only thing to watch out for in future might be to have some check
> against the expected local version number in simtk.openmm.version, just to
> make sure the locally installed version of openmm is as expected.

Good call. I just set up the environments for the current run, so they are fine, but yeah, there should be a check there.

**Comment by jchodera** [comment]
Curious if this is working as expected!

**Comment by pgrinaway** [comment]
It's certainly going much better. I have a couple thousand samples from a ~15 hour run on 96 cores (though @kyleabeauchamp has pointed out there may be additional straightforward optimizations). I need to add in the version check + then we can merge this.

**Comment by pgrinaway** [comment]
Ok, this might be a stupid question, but I'm not quite sure what's going on here. I had thought that sampling concluded normally after running, but instead of 10,000 samples I had only ~3,500.

My question is this: what exactly is an `iteration` to pymc? Profiling the code after I requested 10 iterations with 10 molecules, I see that the `compute_hydration_energies_celery` function is called 141 times. I assumed that this has something to do with the way that pymc makes proposals, though [its docs say](https://pymc-devs.github.io/pymc/modelfitting.html#metropolis-step-methods)

> If a Metropolis step method handles an array-valued variable, it proposes all elements independently but simultaneously.

so that should still be just one update, right? I'm just curious why the likelihood would be apparently called 141 times for only 10 iterations--I assume I'm missing something obvious here.

**Comment by pgrinaway** [comment]
Thinning default, perhaps.

**Comment by pgrinaway** [comment]
That would be a negative, since setting `thin=1` in the sampler constructor still results in 145 evaluations of `compute_hydration_energies_celery` for 10 iterations.

**Comment by kyleabeauchamp** [comment]
Is there an easy way to test a short run with a single celery worker?

**Comment by pgrinaway** [comment]
I can do that, but I don't think the problem is on celery's end, since `compute_hydration_energies_celery` is a wrapper around a celery call.

**Comment by pgrinaway** [comment]
Also, thanks. hope SF is treating you well!

**Comment by jchodera** [comment]
> I'm just curious why the likelihood would be apparently called 141 times for only 10 iterations--I assume I'm missing something obvious here.

Other people are also trying to figure this out:
http://stackoverflow.com/questions/27714635/rationale-for-number-of-pymc-function-evaluations

**Comment by jchodera** [comment]
If array-based parallelization is not working well because it has to, say, re-evaluate a bunch of times if only some element proposals are accepted, we might need to use a different parallelization strategy:
http://continuum.io/blog/bayesian-estimation-hoops

**Comment by jchodera** [comment]
From [an IPython notebook linked to](http://nbviewer.ipython.org/gist/aflaxman/c07c5261bf22f6847098) from [that stackoverflow post](http://stackoverflow.com/questions/27714635/rationale-for-number-of-pymc-function-evaluations), it looks like manually assigning/changing the step methods can reduce the number of energy evaluations.

**Comment by pgrinaway** [comment]
Yeah, I saw that, will give it a try. Right now I'm trying to figure out what it's actually doing, because the bit from the docs

> If a Metropolis step method handles an array-valued variable, it proposes all elements independently but simultaneously.

does not appear to be correct (the proposal is not simultaneous).

**Comment by pgrinaway** [comment]
If you're only updating one variable at a time, don't you have to accept/reject after each attempt? Then there should be NxP samples, no? Or is that not correct?

**Comment by jchodera** [comment]
Huh.  It _looks_ like `MetropolisStep` just makes a single proposal:
https://github.com/pymc-devs/pymc/blob/master/pymc/StepMethods.py#L480-L549

**Comment by pgrinaway** [comment]
Hm, interesting. One thing I noticed was that by adding a `print(parameters)` call to `compute_hydration_energies_celery` was that it was proposing one parameter at a time (as the comment in the actual sampler code [seems to suggest](https://github.com/pymc-devs/pymc/blob/master/pymc/StepMethods.py#L356-L361)). Additionally, I increased the verbosity on the sampler, and see things like this:

```
Metropolis_S_scalingFactor getting initial logp.
    Metropolis_S_scalingFactor Current log-likelihood plus current log-probability -55129393.807
Metropolis_S_scalingFactor proposing.
Metropolis_S_scalingFactor rejecting due to ZeroProbability.
Metropolis_S_scalingFactor returning.
Step method Metropolis_log_sigma stepping
```

many, many more times than 10. I guess my question at this point is, don't these rejections have to be recorded? Should I ask about this on the pymc issue tracker? I'm just a bit confused as to what's going on at this point, so I don't mean to annoy with this.

**Comment by jchodera** [comment]
Might as well ask on the pymc issue tracker if none of us can figure out what is going on here.

It could be that the behavior in the tuning phase is different from the production phase, but I honestly have no idea---presumably someone should document this somewhere.

The log likelihood is really very low.  I wonder why it's so disastrously bad.  Just bad initial parameters for some GB models?

**Comment by jchodera** [comment]
I wonder if the disastrously bad log likelihood suggests we should find the MAP estimate before starting posterior sampling...

**Comment by kyleabeauchamp** [comment]
Worth a try if it converges quickly

**Comment by pgrinaway** [comment]
> It could be that the behavior in the tuning phase is different from the production phase, but I honestly have no idea---presumably someone should document this somewhere.

Good point. I'll play with the options to see if I can vary the tuning.

> The log likelihood is really very low. I wonder why it's so disastrously bad. Just bad initial parameters for some GB models?

Yeah. A lot of atom types have ridiculous "default" parameters (taken from the paper/sander/openmm). For instance, `Br_radius` is only 1.5A. Within a few iterations, it pops up to >3A. I'll try a MAP estimate and see how that goes.

**Comment by pgrinaway** [comment]
I've posted a [question](https://github.com/pymc-devs/pymc/issues/38) on the PyMC issue tracker to request clarification as to how the updates are actually carried out.

**Comment by pgrinaway** [comment]
Ok, so, we have an answer. It looks like the sampler does one-at-a-time Metropolis, then adds the sample to the trace of the variable being sampled. It repeats this `n_parameters` times for each iteration, resulting in `n_iterations` samples for each parameter.

**Comment by jchodera** [comment]
OK, so it essentially does a "sweep" through all parameters each iteration.

**Comment by jchodera** [comment]
I think this is probably OK to merge.  Further optimizations can be made separately.

**Comment by pgrinaway** [comment]
Yeah that sounds good. This is getting a bit out of scope.

**Pull request merged**